subxt/src/blocks/blocks_client.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use crate::{
+    client::OnlineClientT,
+    error::Error,
+    rpc::Rpc,
+    utils::PhantomDataSendSync,
+    Config,
+};
+use derivative::Derivative;
+use futures::{
+    future::Either,
+    stream,
+    Stream,
+    StreamExt,
+};
+use sp_runtime::traits::Header;
+use std::future::Future;
+
+/// A client for working with blocks.
+#[derive(Derivative)]
+#[derivative(Clone(bound = "Client: Clone"))]
+pub struct BlocksClient<T, Client> {
+    client: Client,
+    _marker: PhantomDataSendSync<T>,
+}
+
+impl<T, Client> BlocksClient<T, Client> {
+    /// Create a new [`BlocksClient`].
+    pub fn new(client: Client) -> Self {
+        Self {
+            client,
+            _marker: PhantomDataSendSync::new(),
+        }
+    }
+}
+
+impl<T, Client> BlocksClient<T, Client>
+where
+    T: Config,
+    Client: OnlineClientT<T>,
+{
+    /// Subscribe to new best block headers.
+    ///
+    /// # Note
+    ///
+    /// The best block is selected by the consensus algorithm.
+    ///
+    /// These blocks haven't necessarily been finalised yet. Prefer
+    /// [`BlocksClient::subscribe_finalized_headers()`] if that is important.
+    pub fn subscribe_headers(
```

Comment:
this can't be an async fn because of some lifetime issue or what's it the reason having the the `-> impl Future<_>`?

Reply:
The reason it's used in events, and I think useful here, is so as to not tie the thing you get back to the lifetime of `self`. This just gives it more flexibility to be moved around and such.